### PR TITLE
[FIX] hr_attendance: generated unique kiosk key for every company

### PR DESCRIPTION
**Pull request by sabr-odoo**
This issue occurs when a customer creates multiple companies. Then, after installing the attendance module, after selecting the kiosk mode, an error is generated. This is because, before installing the attendance module, if the customer created a company, a single kiosk key is generated. This key is different for each company, but it is the same.

step to reproduce-
- First created two company.
- Install the `Attendances` & Open it.
- Click on the Kiosk Mode from Menu.
- Error will be generated.

sentry traceback-
```
ValueError: Expected singleton: res.company(2, 1, 4, 3)
  File "odoo/http.py", line 2157, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1732, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1759, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1873, in dispatch
    return self.request.registry['ir.http']._dispatch(endpoint)
  File "addons/website/models/ir_http.py", line 235, in _dispatch
    response = super()._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 207, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 722, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/hr_attendance/controllers/main.py", line 82, in open_kiosk_mode
    } for e in request.env['hr.employee'].sudo().search_read(domain=[('company_id', '=', company.id)],
  File "odoo/fields.py", line 5118, in __get__
    raise ValueError("Expected singleton: %s" % record)
```

after this commit,Each attendance kiosk key generated during the company's creation is unique.

sentry-4617967282

### CODE REVIEW

**Review by thle-odoo — APPROVED** [review]
Hello @sabr-odoo,

In fact, even if the `attendance_kiosk_key` field has a lambda as its default value, it will only be calculated once and this value will be apply in DB for all the records (the companies in this case) [query](https://github.com/odoo/odoo/blob/17.0/odoo/models.py#L3058).

It's OK to go through all the records again and regenerate the token as there aren't too many companies.

However, there is another way to set a different token on existing records, by overriding `_init_column` and setting the token directly via a query.
Here's the link where we use this method: https://github.com/odoo/odoo/blob/9ebbfdac7378ac0e16eac9e08175a4143fb50061/addons/website_slides/models/slide_channel.py#L702-L716

The final reviewer will decide which of these methods to use in this case.

Maybe add a `limit=1` in the `search` of the `_get_company` method, as this method should only return one company (but that's just a detail).

Thanks for your work !

**Comment by sabr-odoo** [comment]
Hello @thle-odoo ,

I think `limit=1` is not good because for example the user has five companies, which are `A,B,C,D,E` and is currently in C company, try to open the kiosk mode. At that time, the user will get the A company in res.company(); however, the ``attendance_kiosk_key` field must be unique.

And also, changes are done by overriding  `_init_column`

**Comment by jugj-odoo** [comment]
Hello @sabr-odoo and @thle-odoo , little consistency issue, the init generates a key of the format xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx, while the regenerate is in hex format, so we should pick one format and apply it to both

**Review by sofiagvaladze — CHANGES_REQUESTED** [review]
@sabr-odoo thanks for your work :raised_hands: 

 some minor cosmetic changes, otherwise code LGTM

File: addons/hr_attendance/models/res_company.py [hunk 1]
```diff
@@ -96,3 +96,23 @@ def _action_open_kiosk_mode(self):
...
+    # ORM Overrides
+    # ---------------------------------------------------------
+
+    def _init_column(self, column_name):
+        """ Initialize the value of the given column for existing rows.
+            Overridden here because we need to generate different access tokens
+            and by default _init_column calls the default method once and applies
+            it for every record.
+        """
+        if column_name != 'attendance_kiosk_key':
```
[the remark this answers is not in the document]
Comment (sabr-odoo) [reply addons/hr_attendance/models/res_company.py]: Hello :wave: @sofiagvaladze 

The `Attendance` module installation error might have caused these changes to be not good.
 
```
File "/home/odoo/odoo/community/odoo/sql_db.py", line 332, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.UndefinedColumn: column "attendance_kiosk_key" does not exist
LINE 4:                 WHERE attendance_kiosk_key IS NULL
```

[the remark this answers is not in the document]
Comment (sofiagvaladze) [reply addons/hr_attendance/models/res_company.py]: What I mean, was changing the placing of blocks.. not just if statement.. obviously that won't work

**Comment by sofiagvaladze** [comment]
@odoo/rd-security can you check this PR?

**Comment by mart-e** [comment]
@sofiagvaladze the md5 is just to have something unique? Why not a uuid4() like everywhere else?

**Comment by sofiagvaladze** [comment]
@mart-e indeed, that's what we want. We were also thinking about passing the list of touples, like it is done in - 

https://github.com/odoo/enterprise/blob/4b66e42b47656192c8f4d6787a13607e37950457/planning/models/hr.py#L48

So, we will do that to avoid using md5 and all that. 

@sabr-odoo, what do you think? Again, we can use the link above for inspiration.

**Comment by sabr-odoo** [comment]
> @mart-e indeed, that's what we want. We were also thinking about passing the list of touples, like it is done in -
> 
> https://github.com/odoo/enterprise/blob/4b66e42b47656192c8f4d6787a13607e37950457/planning/models/hr.py#L48
> 
> So, we will do that to avoid using md5 and all that.
> 
> @sabr-odoo, what do you think? Again, we can use the link above for inspiration.

Hello :wave:  @sofiagvaladze , changes are done

File: addons/hr_attendance/models/res_company.py [hunk 2]
```diff
@@ -36,6 +39,29 @@ def _compute_attendance_kiosk_url(self):
...
+            super(ResCompany, self)._init_column(column_name)
+        else:
+            self.env.cr.execute("SELECT id FROM %s WHERE attendance_kiosk_key IS NULL" % self._table)
+            acc_ids = self.env.cr.dictfetchall()
+            values_args = [(acc_id['id'], self._default_employee_token()) for acc_id in acc_ids]
```
Comment (sofiagvaladze) [inline addons/hr_attendance/models/res_company.py:56]: Blind copy pasting is never a good idea. 
what does acc_ids stand for here?

```suggestion
            attendance_ids = self.env.cr.dictfetchall()
            values_args = [(attendance_id['id'], self._default_company_token()) for attendance_id in attendance_ids]
```

File: addons/hr_attendance/models/res_company.py [hunk 3]
```diff
@@ -10,6 +10,9 @@
 class ResCompany(models.Model):
     _inherit = 'res.company'
 
+    def _default_employee_token(self):
```
Comment (sofiagvaladze) [inline addons/hr_attendance/models/res_company.py:13]: Why employee? :smile: I am not against copying the code and using it for inspiration, but then we need to adjust it, no? 

```suggestion
    def _default_company_token(self):
```

**Review by sofiagvaladze — CHANGES_REQUESTED** [review]

**Comment by sabr-odoo** [comment]
@sofiagvaladze ,changes are done

**Comment by sofiagvaladze** [comment]
robodoo r+

**Pull request closed without merging**